File: app/src/main/java/com/example/test/mvvmsampleapp/di/AppModule.java
package com.example.test.mvvmsampleapp.di;

import android.arch.lifecycle.ViewModelProvider;

import com.example.test.mvvmsampleapp.service.repository.GitHubService;
import com.example.test.mvvmsampleapp.viewmodel.ProjectViewModelFactory;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module(subcomponents = ViewModelSubComponent.class)
class AppModule {
    @Singleton @Provides
    GitHubService provideGithubService() {
        return new Retrofit.Builder()
                .baseUrl(GitHubService.HTTPS_API_GITHUB_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(GitHubService.class);
    }

    @Singleton
    @Provides
    ViewModelProvider.Factory provideViewModelFactory(
            ViewModelSubComponent.Builder viewModelSubComponent) {

        return new ProjectViewModelFactory(viewModelSubComponent.build());
    }
}

File: app/src/main/java/com/example/test/mvvmsampleapp/di/AppModule.kt
package com.example.test.mvvmsampleapp.di

import android.arch.lifecycle.ViewModelProvider
import com.example.test.mvvmsampleapp.service.repository.GitHubService
import com.example.test.mvvmsampleapp.viewmodel.ProjectViewModelFactory
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module(subcomponents = [(ViewModelSubComponent::class)])
internal class AppModule {

    @Singleton
    @Provides
    fun provideGithubService(): GitHubService {
        return Retrofit.Builder()
                .baseUrl(GitHubService.HTTPS_API_GITHUB_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(GitHubService::class.java)
    }

    @Singleton
    @Provides
    fun provideViewModelFactory(
            viewModelSubComponent: ViewModelSubComponent.Builder): ViewModelProvider.Factory {

        return ProjectViewModelFactory(viewModelSubComponent.build())
    }
}
